### PR TITLE
[Improvement](local shuffle) Improve local shuffle strategy

### PR DESCRIPTION
**Pull request by Gabriel39**
## Proposed changes

Add local shuffle to unpartitioned fragment to add parallel for perfomance
```sql
SELECT h1.UserID, h2.URL, COUNT(*) AS visit_count
FROM (
  SELECT *
  FROM hits_10m
  LIMIT 5000
) AS h1
CROSS JOIN (
  SELECT *
  FROM hits_10m
  LIMIT 5000
) AS h2
GROUP BY h1.UserID, h2.URL
ORDER BY visit_count DESC
LIMIT 1000
```

Add a rule to apply local exchanger:

```
┌───────────────────────┐               ┌───────────────────────┐
│                       │               │                       │
│Exchange(UNPARTITIONED)│               │Exchange(UNPARTITIONED)│
│                       │               │                       │
└───────────────────────┴──────┬────────┴───────────────────────┘
                               │                                 
                               │                                 
                               │                                 
                               │                                 
                               │                                 
                               │                                 
                        ┌──────▼──────┐                          
                        │             │                          
                        │ CROSS JOIN  │                          
                        │             │                          
                        └──────┬──────┘                          
                               │                                 
                               │                                 
                               │                                 
            ┌──────────────────▼─────────────────────┐           
            │                                        │           
            │ LOCAL EXCHANGE (HASH PARTITION) 1 -> n │           
            │                                        │           
            └──────────────────┬─────────────────────┘           
                               │                                 
                               │                                 
                               │                                 
                               │                                 
                   ▼        ┌──▼────┐                            
                            │       │                            
                            │  AGG  │                            
                            │       │                            
                            └───────┘                            
```

before: 1 min 17.79 sec
after: 16.73 sec

### CODE REVIEW

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Review by morrySnow — COMMENTED** [review]
please provide a detailed description of the purpose and reasons for the modifications.

File: fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java [hunk 1]
```diff
@@ -1779,6 +1779,22 @@ protected void computeFragmentHosts() throws Exception {
                 FInstanceExecParam instanceParam = new FInstanceExecParam(null, execHostport,
                         0, params);
                 params.instanceExecParams.add(instanceParam);
+
+                int expectedInstanceNum = fragment.getParallelExecNum();
+                boolean forceToLocalShuffle = context != null
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java:1784]: could u add some comment to explain what do u want to do

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Review by 924060929 — DISMISSED** [review]

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Review by 924060929 — DISMISSED** [review]

**Review by BiteTheDDDDt — DISMISSED** [review]

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

**Review by BiteTheDDDDt — APPROVED** [review]

**Review by 924060929 — APPROVED** [review]

**Pull request merged**